tests/test_proxy_connect.py

```diff
@@ -17,31 +18,42 @@
 
 class HTTPSProxy(controller.Master, Thread):
 
-    def __init__(self, port):
+    def __init__(self):
         password_manager = http_auth.PassManSingleUser('scrapy', 'scrapy')
         authenticator = http_auth.BasicProxyAuth(password_manager, "mitmproxy")
         cert_path = os.path.join(os.path.abspath(os.path.dirname(__file__)),
             'keys', 'mitmproxy-ca.pem')
         server = proxy.ProxyServer(proxy.ProxyConfig(
             authenticator = authenticator,
             cacert = cert_path),
-            port)
+            0)
+        self.server = server
         Thread.__init__(self)
         controller.Master.__init__(self, server)
 
+    def http_address(self):
+        return 'http://scrapy:scrapy@%s:%d' % self.server.socket.getsockname()
+
+
+def _wrong_credentials(proxy_url):
+    bad_auth_proxy = list(urlsplit(proxy_url))
+    bad_auth_proxy[1] = bad_auth_proxy[1].replace('scrapy:scrapy@', 'wrong:wronger@')
+    return urlunsplit(bad_auth_proxy)
 
 class ProxyConnectTestCase(TestCase):
 
     def setUp(self):
         self.mockserver = MockServer()
         self.mockserver.__enter__()
         self._oldenv = os.environ.copy()
-        self._proxy = HTTPSProxy(8888)
+
+        self._proxy = HTTPSProxy()
         self._proxy.start()
+
         # Wait for the proxy to start.
         time.sleep(1.0)
-        os.environ['http_proxy'] = 'http://scrapy:scrapy@localhost:8888'
-        os.environ['https_proxy'] = 'http://scrapy:scrapy@localhost:8888'
+        os.environ['https_proxy'] = self._proxy.http_address()
```

Comment:
should it be ``self._proxy.https_address()`` instead?

Reply:
We have two type of protocols - protocol we talk with a crawled server and protocol we talk with a proxy. For different protocols for crawled server(http and https) we use same proxy protocol(http).